File: sdk/python/examples/langchain/02_react_with_tools.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""ReAct Agent with Tools — agent with practical utility tools.

Demonstrates:
    - Defining tools with @tool decorator
    - Passing tools to create_agent for a ReAct-style loop
    - Calculator, string, and date utilities

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:6767/api
    - OPENAI_API_KEY for ChatOpenAI
"""

import math
from datetime import date

from langchain.agents import create_agent
from langchain_core.tools import tool
from langchain_openai import ChatOpenAI
from agentspan.agents import AgentRuntime


@tool
def calculate(expression: str) -> str:
    """Evaluate a safe mathematical expression and return the result.

    Supports +, -, *, /, **, sqrt, pi. Example: 'sqrt(144)', '2 ** 10'
    """
    try:
        result = eval(expression, {"__builtins__": {}}, {"sqrt": math.sqrt, "pi": math.pi})
        return str(result)
    except Exception as e:
        return f"Error: {e}"


@tool
def count_words(text: str) -> str:
    """Count the number of words in the provided text."""
    return f"{len(text.split())} words"


@tool
def get_today() -> str:
    """Return today's date in YYYY-MM-DD format."""
    return date.today().isoformat()


llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)

tools = [calculate, count_words, get_today]

graph = create_agent(llm, tools=tools, name="react_tools_agent")

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(
            graph,
            "What is sqrt(256)? Also count words in 'the quick brown fox'. What is today's date?",
        )
        print(f"Status: {result.status}")
        result.print_result()

        # Production pattern:
        # 1. Deploy once during CI/CD:
        # runtime.deploy(graph)
        # CLI alternative:
        # agentspan deploy --package examples.langchain.02_react_with_tools
        #
        # 2. In a separate long-lived worker process:
        # runtime.serve(graph)

File: sdk/typescript/examples/langgraph/02-react-with-tools.ts
/**
 * ReAct Agent with Tools -- createReactAgent with practical tools.
 *
 * Demonstrates:
 *   - Defining tools with DynamicStructuredTool from @langchain/core/tools
 *   - Passing tools to createReactAgent for a ReAct-style loop
 *   - Multi-tool invocation in a single query
 */

import { createReactAgent } from '@langchain/langgraph/prebuilt';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage } from '@langchain/core/messages';
import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { AgentRuntime } from '../../src/index.js';

// ---------------------------------------------------------------------------
// Tool definitions
// ---------------------------------------------------------------------------
const calculateTool = new DynamicStructuredTool({
  name: 'calculate',
  description: 'Evaluate a mathematical expression. Supports basic arithmetic, sqrt, and pi.',
  schema: z.object({
    expression: z.string().describe('The mathematical expression to evaluate'),
  }),
  func: async ({ expression }) => {
    try {
      const sanitized = expression.replace(/[^0-9+\-*/().sqrtpi ]/g, '');
      const result = Function(
        '"use strict"; const sqrt = Math.sqrt; const pi = Math.PI; return (' +
          sanitized +
          ')',
      )();
      return String(result);
    } catch (e) {
      return `Error evaluating expression: ${e}`;
    }
  },
});

const countWordsTool = new DynamicStructuredTool({
  name: 'count_words',
  description: 'Count the number of words in a given text.',
  schema: z.object({
    text: z.string().describe('The text to count words in'),
  }),
  func: async ({ text }) => {
    const words = text.trim().split(/\s+/);
    return `The text contains ${words.length} word(s).`;
  },
});

const getTodayTool = new DynamicStructuredTool({
  name: 'get_today',
  description: "Return today's date in YYYY-MM-DD format.",
  schema: z.object({}),
  func: async () => new Date().toISOString().slice(0, 10),
});

// ---------------------------------------------------------------------------
// Build the graph
// ---------------------------------------------------------------------------
const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });
const graph = createReactAgent({
  llm,
  tools: [calculateTool, countWordsTool, getTodayTool],
});

const PROMPT =
  "What is the square root of 256? Also, how many words are in 'the quick brown fox'? And what is today's date?";

// ---------------------------------------------------------------------------
// Run
// ---------------------------------------------------------------------------
async function main() {
  // ── Path 1: Native ──
  console.log('=== Native LangGraph execution ===');
  const nativeResult = await graph.invoke({
    messages: [new HumanMessage(PROMPT)],
  });
  for (const msg of nativeResult.messages) {
    const role = msg.constructor.name;
    if (msg.tool_calls && msg.tool_calls.length > 0) {
      console.log(`  ${role}: [tool_calls]`, JSON.stringify(msg.tool_calls.map((tc: any) => tc.name)));
    } else {
      const content = typeof msg.content === 'string' ? msg.content.slice(0, 200) : JSON.stringify(msg.content);
      console.log(`  ${role}: ${content}`);
    }
  }

  // ── Path 2: Agentspan ──
  console.log('\n=== Agentspan passthrough execution ===');
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(graph, PROMPT);
    console.log('Status:', result.status);
    result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

main().catch(console.error);
